ENH/DEP: Use a ufunc under the hood for ndarray.clip

eric-wieser opened the pull request:
This includes:
* The addition of 3-input PyObject inner loop
* The removal of `->f->fastclip` for builtin types, which now use ufuncs instead
* A deprecation in `PyArray_Clip` for third-party types that still use `->f->fastclip`
* A deprecation of the unusual casting behavior of `clip`
* A deprecation of the broken `nan`-behavior of `clip`, which was previously dependent on dimensionality and byte-order.
* **new:** A behavior change for `max < min` which brings object arrays in line with the other types

Surprisingly, this actually seems to have better performance (N=10000):
```python
In [1]: a = np.arange(10000, dtype=float)
In [2]: a2 = np.empty(a.shape, dtype=[('a', a.dtype), ('b', np.bool)])['a']
In [3]: a2[...] = a

In [4]: %timeit np.clip(a, 0, 1)
20.3 µs ± 206 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit np.core.umath.clip(a, 0, 1)
16.9 µs ± 1.37 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit np.clip(a2, 0, 1)
65.7 µs ± 1.11 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [7]: %timeit np.core.umath.clip(a2, 0, 1)
62.6 µs ± 970 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Edit: those performance numbers are probably not representative of the wrapped call - there is a big cost to pay for the deprecations.

eric-wieser committed:
ENH/DEP: Use a ufunc under the hood for ndarray.clip

This includes:
* The addition of 3-input PyObject inner loop
* The removal of `->f->fastclip` for builtin types, which now use ufuncs instead
* A deprecation in `PyArray_Clip` for third-party types that still use `->f->fastclip`
* A deprecation of the unusual casting behavior of `clip`
* A deprecation of the broken `nan`-behavior of `clip`, which was previously dependent on dimensionality and byte-order.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -55,6 +55,18 @@ def _count_reduce_items(arr, axis):
+        return um.fmin(a, a_max, out=out)
+    elif a_max is None:
+        # note: nan-behaviour is incorrect here
+        return um.fmax(a, a_min, out=out)
```
* max: returns `nan` if either input is NaN
* fmax: returns the non-nan argument if either input is nan
* What we need here: returns the first argument if either input is nan

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2921,6 +2930,95 @@ OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
+ *****************************************************************************
+ **                               FASTCLIP                                  **
+ *****************************************************************************
+ */
```
I really think we want to split the ufunc loops into multiple files, possible one per ufunc. Keeping track of where to insert things is becoming extremely hard.

Unfortunately, right after we branch the 1.16 LTS(ish?) release is a terrible time to make the history diverge like that.

eric-wieser replied:
Moved just `clip` to its own file for now

eric-wieser replied:
And the macros are now separate too, in #13032

mattip commented:
if you have a ready benchmark, perhaps you could add it?

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2921,6 +2930,95 @@ OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
+
+/*
+ * In fastclip, 'b' was already checked for NaN, so the half comparison
+ * only needs to check 'a' for NaN.
```
I think this comment might now be false.

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2921,6 +2930,95 @@ OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
+        npy_intp n = dimensions[0];
+        npy_intp i;
+
+        /* contiguous, branch to let the compiler optimize */
```
Just enough optimization here to make `um.clip` faster than `np.clip` in the cases that are common.

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2921,6 +2930,95 @@ OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
+ *
+ * #name = CFLOAT, CDOUBLE, CLONGDOUBLE#
+ * #type = npy_cfloat, npy_cdouble, npy_clongdouble#
+ */
```
To be removed...

eric-wieser commented:
@mattip: Benchmark added. I don't know how much energy I have to put into this, but this looks like the lowest-hanging fruit of #12514. Feel free to take over this PR as your own, if you want to.

eric-wieser commented:
Outdated but non-resolved comments above still apply

eric-wieser commented on numpy/core/code_generators/generate_umath.py:
```diff
@@ -502,6 +502,24 @@ def english_upper(s):
+          'PyUFunc_DefaultTypeResolver',
+          TD(noobj),
+          ),
+'clip_below':
```
Naming is hard. I indended this to mean "clip everything _that is_ below a certain value", but it could be equally read as "clip everything _to_ below a certain value".

Naming advice?

eric-wieser replied:
Now moot, @mhvk argues these overloads provided no value.

mhvk commented on numpy/core/_methods.py:
```diff
@@ -55,6 +55,18 @@ def _count_reduce_items(arr, axis):
+    if min is None and max is None:
+        raise ValueError("One of max or min must be given")
+    elif min is None:
+        return um.clip_above(a, max, out=out, casting=casting)
```
Can we not use `um.minimum` here? (and `um.maximum` instead of `clip_below`)

eric-wieser replied:
Nope - clip needs to preserve NaNs in and only in the first argument. maximum always propagates, and fmax always ignores. So we need a third maximum-like function

mhvk commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -39,7 +39,12 @@ def get(name):
+    )
+
+    if name[0] != '_' and name not in skip:
         # matmul is special, it does not use the OUT_SCALAR replacement strings
```
Need to adjust the comment here on why `clip` is excluded; though perhaps better to put it behind each item in `skip`.

mhvk commented:
Taking the conversation of using `minimum` and `maximum` for `clip` outside of the in-line comments (since I missed the first one...): is it obvious what the behaviour should be? I think that if `min` or `max` is `NaN`, it is not strange to just pass that on. Oddly, that is *partially* what the current `clip` does:
```
a = np.array([10., np.nan, np.nan, -10.])
a.clip(np.array([np.nan, np.nan, 1., 1.]), np.array([2., 2., np.nan, np.nan]))
# RuntimeWarning: invalid value encountered in minimum
# RuntimeWarning: invalid value encountered in maximum
# array([nan, nan, nan, nan])
```
Looking at the code, I see this is because for array-like min and max, the `_slow_array_clip` path is chosen, which in fact uses the `minimum` and `maximum` ufuncs!

My sense would be to *keep* this behaviour, following the maxim that in case of doubt one should refuse the temptation to guess: the user should be `±np.inf` if a particular element should not have a minimum or maximum.

Note that if we felt it was needed, we could keep this completely backwards compatible if in the wrapping code we checked for cases of scalar `min/max=np.nan` and treated those as equivalent `None` and then also raised a `DeprecationWarning`.

eric-wieser commented:
> is it obvious what the behaviour should be

Not really - I just noticed that in the past someone complained when the `nan` was propagated (#7601)

> the user should be ±np.inf if a particular element should not have a minimum or maximum.

That's a fairly strong argument I hadn't thought of as to why the nan behavior is useless. Unfortunately the replacement is quite a bit more verbose.
```
np.clip(arr, np.where(np.isnan(a_min), -np.inf, a_min), np.where(np.isnan(a_max), np.inf, a_max))
```

mhvk commented:
@eric-wieser - #7601 complained that `nan` in the array was *not* propagated on some windows virtual machine! I think we agree that no matter what, `nan` in the input array should be propagated (i.e., we will not use `fmin`, `fmax` behaviour). The question is what to do with `nan` in min/max.

Here, remember that at present `nan` *do* get propagated if `min` or `max` are arrays (since at present `minimum` and `maximum` are used for that case). So, the question is whether we want to extend the scalar behaviour of ignoring min,max=nan to arrays or extend the array behaviour of propagating min,max=nan to scalars (possibly with deprecation).

My own sense is to always propagate `nan` for the reasons I gave earlier; it also may be the least likely to lead to regressions, since having a `nan` in an array of `min` or `max` seems much more likely than passing in a `nan` scalar for `min` or `max`. (But then, scalars are surely used much more often... Which perhaps argues for deprecation of the `nan` scalar case....)

mhvk commented:
OK, makes sense to just have it in the release notes. So, we change to always behave consistently with `mininum(maximum(array, clip_low), clip_high)`?

eric-wieser commented:
Finally back out of the rabbit hole - this was not an easy conversion.

> So, we change to always behave consistently with `mininum(maximum(array, clip_low), clip_high)`?

Essentially yes, but I've bent over backwards to try and deprecate the old behavior rather than outright breaking it.

This will break custom types that implement `fastclip`, but I suspect there are exactly 0 of those in the wild.

mhvk commented on numpy/core/src/umath/clip.c.src:
```diff
@@ -0,0 +1,119 @@
+        npy_intp is1 = steps[0], os1 = steps[3];
+        npy_intp n = dimensions[0];
+
+        /* contiguous, branch to let the compiler optimize */
```
Compilers are smart enough for that? To help the poor humans, I'd perhaps have done `@type@ *_op1 = (@type@ *)op1` and then `_op1++` in the loop to make it more obvious... But as long as you're sure this works, it is definitely fine by me.

eric-wieser replied:
This matches the approach used by other loops. It's possible an alignment check is needed here too though.

mhvk replied:
How do we find out? If we're not sure, we should at least note that.

mhvk commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4264,7 +4264,7 @@ def _clip_type(self, type_group, array_max,
                 x = (np.random.random(1000) * array_max).astype(dtype)
                 if inplace:
-                    x.clip(clip_min, clip_max, x)
+                    x.clip(clip_min, clip_max, x, casting='unsafe')
```
Is this still necessary with the change from int to float that you made below?

eric-wieser replied:
Yes, this still causes problems on signed / unsigned integer promotion

mhvk replied:
Maybe add a comment...

mhvk commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1480,7 +1480,9 @@ def test_simple_int32_inout(self):
-        self.fastclip(a, m, M, ac)
+        with assert_warns(DeprecationWarning):
+            # NumPy 1.17.0, 2018-02-24 - casting is unsafe
+            self.fastclip(a, m, M, ac)
```
Would it make sense to add the `casting='unsafe'` version and check it gives no warnings?

eric-wieser replied:
Probably, but I ran out of energy for more tests here

tylerjereddy replied:
I've parametrized this unit test to check for both `casting=None` and `casting="unsafe"` scenarios, as requested. As expected, the latter does indeed suppress the warning.

mhvk reviewed (commented):
Looks good overall. But compilation fails. And of course some comments!

eric-wieser commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -2528,6 +2533,36 @@ def add_newdoc(place, name, doc):
+        Array containing elements to clip.
+    a_min : array_like
+        Minimum value.
+    a_max : scalar or array_like or `None`
```
None has no special meaning for the ufunc, it implies an object array and probably a TypeError

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -1938,12 +1938,12 @@ def compress(condition, a, axis=None, out=None):
 
 
-def _clip_dispatcher(a, a_min, a_max, out=None):
+def _clip_dispatcher(a, a_min, a_max, out=None, **kwargs):
```
I agree - what's left of clip is just a ufunc selector based on argument presence, there's no reason to allow customization of that.

shoyer replied:
Well, the current promise is that *everything* directly exposed by NumPy [gets its own dispatcher](http://www.numpy.org/neps/nep-0018-array-function-protocol.html#implementations-in-terms-of-a-limited-core-api), either via `__array_function__` or `__array_ufunc__`. This PR means that `np.clip` now uses one of three ufuncs, one of which (`um.clip`) isn't even publicly exposed. So I don't think we can get rid of this dispatcher yet.

eric-wieser replied:
Should we consider `np.core.umath.clip` sufficient exposure? It can never replace `np.clip` in the top namespace and IMO there's no clearer name available

mhvk replied:
I'd think that `np.core.umath` is sufficient exposure (indeed, in astropy we check our ufunc coverage by enumerating what is in there). The larger question might be whether this multi-dispatch would still need an override. To me, that seems a bit odd, and counter to the idea that `__array_function__` is only meant to cover what isn't covered by better matched overrides.

p.s.  An analogy here may be `ndarra.__pow__`, which gives different `ufuncs` depending on the argument (though of course subclasses can override that...).

shoyer replied:
The rest of `np.core` is not considered public API so I don't think that's enough exposure. Maybe we could put it as an attribute on the function, e.g., `np.clip.ufunc`?

I'm also reluctant to do this on an ad-hoc basis. I don't think it's worth the complexity of having exactly *one* special function that is neither a ufunc nor with a direct override via `__array_function__`. We don't need to change this all at once but at least we should have a rule for going forward that clearly applies to more than only a few functions.

mhvk replied:
OK, perhaps not that simple. Maybe for now best to finish this PR without removing the `__array_function__` override?

eric-wieser commented on numpy/core/src/multiarray/calculation.c:
```diff
@@ -918,6 +918,27 @@ PyArray_Clip(PyArrayObject *self, PyObject *min, PyObject *max, PyArrayObject *o
+            return PyObject_CallFunctionObjArgs(n_ops.maximum, self, min, out, NULL);
+        }
+        else {
+            return PyObject_CallFunctionObjArgs(n_ops.conjugate, self, min, max, out, NULL);
```
Oops. As you might guess, this function has no tests now, as it has been removed from any python code paths

eric-wieser commented on numpy/core/src/umath/clip.c.src:
```diff
@@ -0,0 +1,119 @@
+#define _NPY_@name@_MAX(a, b) PyArray_MAX(a, b)
+/**end repeat**/
+
+#define _NPY_HALF_MIN(a, b) (npy_half_isnan(a) || npy_half_le(a, b) ? (a) : (b))
```
I think this matches the order the minimum and maximum ufuncs use. I believe swapping them produced an msvc bug. @mattip may remember more.

mhvk replied:
Wouldn't our tests show whether there is such a compiler bug?  Maybe at least add a comment to file that a speed-up may be possible? Or are compilers smart enough to realize that `isnan` is usually an unlikely case and should be done last?

eric-wieser commented on numpy/core/src/umath/clip.c.src:
```diff
@@ -0,0 +1,119 @@
+ */
+#define _NPY_@name@_MIN(a, b) (npy_isnan(a) ? (a) : PyArray_MIN(a, b))
+#define _NPY_@name@_MAX(a, b) (npy_isnan(a) ? (a) : PyArray_MAX(a, b))
+#define _@name@_ISNAN(x) npy_isnan(x)
```
I think this is now unused

mhvk replied:
It is unused so can be removed.

tylerjereddy replied:
I've removed this line since you're both in agreement there

eric-wieser commented:
Not sure I understand [the travis warning](https://travis-ci.org/numpy/numpy/jobs/499678192#L2536)

mhvk commented:
Indeed, strange that the warning wasn't present before...

mhvk commented on numpy/core/src/multiarray/calculation.c:
```diff
@@ -918,6 +918,27 @@ PyArray_Clip(PyArrayObject *self, PyObject *min, PyObject *max, PyArrayObject *o
+    }
+    /* everything below can be removed once this deprecation completes */
+
     if (func == NULL
```
`func` can not be `NULL` here.

mhvk commented on numpy/core/src/umath/clip.c.src:
```diff
@@ -0,0 +1,119 @@
+ * #name = DATETIME, TIMEDELTA#
+ */
+#define _NPY_@name@_MIN(a, b) ( \
+    (a) == NPY_DATETIME_NAT ? (a) : \
```
Here both arguments are tested, but I think only one should be necessary: DATETIME_NAT Is the minimum, so is guaranteed to be the smaller of the two (though perhaps this is overoptimizing...)

tylerjereddy replied:
This I'm not clear on. If it is worth avoiding the standard minimum ternary operation with an early exit when `a` is `NaT`, why wouldn't it be worth the similar early exit if `b` is `NaT`? Seems like a pretty small detail though.

I'm slightly more confused by the `_NPY_@name@_MAX` function just below this being identical to the `MIN` version. If that is correct, can't you just call the MIN version from MAX then? Maybe there's some inner loop / custom template language reason we can't, but seems like `@name@` would just get substituted in & call the other function?

mhvk commented on numpy/core/src/umath/clip.c.src:
```diff
@@ -0,0 +1,119 @@
+@name@_clip(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
+{
+    if (steps[1] == 0 && steps[2] == 0) {
+        /* min and max are constant throughout the loop, the most common case */
```
It would seem that for this case one could optimize the checks for `nan` - maybe for now leave a note to that extent?

tylerjereddy replied:
I've added the requested comment about possible `nan` optimization in the future

eric-wieser replied:
I think the compiler will spot the repeated function invocation anyway, not something we need to worry about.

mhvk commented on numpy/core/src/umath/funcs.inc.src:
```diff
@@ -226,6 +226,17 @@ npy_ObjectLCM(PyObject *i1, PyObject *i2)
+    if (o == NULL) {
+        return NULL;
+    }
+    Py_SETREF(o, npy_ObjectMin(arr, max));
```
This should be `npy_ObjectMin(o, max)`! Also means there is no test yet that catches this...

eric-wieser replied:
Oops!

tylerjereddy replied:
Ok, patched with regression test now, so marking as resolved.

mhvk reviewed (commented):
Mostly little things, but one bug for the object case.

tylerjereddy reviewed (commented):
This currently works differently on master vs. this branch:
 
```python
a = np.arange(10, dtype=object) 
result = np.clip(a, 1, 5)
```

```
result: [1 1 2 3 4 5 5 5 5 5] # master branch
result: [0 1 2 3 4 5 5 5 5 5] # feature branch
```
Is the reason for this obvious, or should I try to debug it? I guess a unit test of that nature might be helpful?

I tried applying the fix Marten suggested for `npy_ObjectClip()` without success. That's what lead me to poking around with `dtype=object` in the first place, but maybe this is just misguided probing.

mhvk commented:
Strange that my fix didn't solve it - certainly this would be a good test to add!

tylerjereddy commented:
Ah, you're right, it does fix it -- I must not have been cleaning between builds carefully enough.

tylerjereddy committed:
MAINT: address reviewer comments

* Unit test for object clip issue

* parametrized test_simple_int32_inout() to
additionally handle scenario where "unsafe"
casting is explicitly passed through, as
requested by reviewer

* add a requested comment related to possibility
of future nan check optimization for `@name@_clip`

* add a unit test to cover the case where np.clip()
is called with None for both max and min

* add a unit test for the case where out is None
and casting is specified as an invalid value, to flush
through code path in TestClip.fastclip

* add unit test + doc update for expected behavior
when amin > amax

* add unit test + patch for bug in npy_ObjectClip; its
operations and operands were wrong; the unit test case
is based on a scenario probed by the hypothesis library

* add unit test for error in timedelta64 MAX
function for clip handling

* add unit test for a pathological case where np.ones(10)
is clipped with amin=1, amax=0

tylerjereddy commented:
I pushed a unit test & the simple fix to my fork on branch `clip-obj`.

The diff if you prefer:

```diff
diff --git a/numpy/core/src/umath/funcs.inc.src b/numpy/core/src/umath/funcs.inc.src
index 475c246c86..b4a19b04cb 100644
--- a/numpy/core/src/umath/funcs.inc.src
+++ b/numpy/core/src/umath/funcs.inc.src
@@ -233,7 +233,7 @@ npy_ObjectClip(PyObject *arr, PyObject *min, PyObject *max) {
     if (o == NULL) {
         return NULL;
     }
-    Py_SETREF(o, npy_ObjectMin(arr, max));
+    Py_SETREF(o, npy_ObjectMin(o, max));
     return o;
 }

diff --git a/numpy/core/tests/test_numeric.py b/numpy/core/tests/test_numeric.py
index d0b8c0d036..7568a6ee13 100644
--- a/numpy/core/tests/test_numeric.py
+++ b/numpy/core/tests/test_numeric.py
@@ -1798,6 +1798,11 @@ class TestClip(object):
         with assert_warns(DeprecationWarning):
             assert_equal(d.clip(min=np.nan, max=10), d)

+    def test_object_clip(self):
+        a = np.arange(10, dtype=object)
+        actual = np.clip(a, 1, 5)
+        expected = np.array([1, 1, 2, 3, 4, 5, 5, 5, 5, 5])
+        assert actual.tolist() == expected.tolist()

 class TestAllclose(object):
     rtol = 1e-5
```

eric-wieser commented:
Feel free to push over my branch, I'll make sure to fetch next time I pick this back up

mhvk commented:
It may be a few days before I have time to look at this, but if @eric-wieser and you agree, just go ahead and merge.

tylerjereddy commented:
Not unique to this feature branch, but I was doing a little property-based testing on `np.clip()` and noticed this happens without a warning:

```python
np.clip(np.zeros(10), 1, 0)
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
```

Not sure if intentional--strange to think about. Original property based test below the fold
<details>

```python
import numpy as np
import hypothesis
from hypothesis import given
from hypothesis import strategies as st
from hypothesis.extra import numpy as hynp

@given(hynp.arrays(np.int64, 10),
       st.integers(),
       st.integers(),
       )
def test_clip(arr, amin, amax):
    result = np.clip(arr, amin, amax)
    # preserve shape on clip
    assert result.shape == arr.shape
    # respect clip limits
    assert result.max() <= amax
    assert result.min() >= amin
```

</details>

tylerjereddy commented:
I've added a patch & unit test for the case described above, based on feedback from the meeting today. Even better would be to add the property-based tests directly, but I'm guessing that is controversial because of the dependency addition for tests (and I'd probably have *way* too much fun writing property-based tests to discover weaknesses).

I'm also not sure how much we want to conflate the migration to the ufunc with fixing bugs / issues, but I've added this patch tentatively anyway.

tylerjereddy commented:
Looks like my check for `min > max` is not Windows portable

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -96,6 +97,10 @@ def _clip_dep_invoke_with_casting(ufunc, *args, out=None, casting=None, **kwargs
+    elif min is not None and max is not None:
+        # need to handle scalar & array-like comparisons here
+        if (np.asanyarray(min) > np.asanyarray(max)).any():
+            raise ValueError("clip a_min greater than a_max")
```
I think setting an `invalid` flag within the ufunc would make more sense here, the same way we do for `/ 0`

mhvk replied:
Do we really want to make the ufunc substantially slower just for this sanity check? To me it makes more sense to define the behaviour and document it.

eric-wieser replied:
I could go either way. We certainly don't want to invoke another ufunc in python, but the cost of another comparison might be marginal

mhvk replied:
Within the ufunc, the cost is probably less, especially for the common case of constant min and max. 

On the other hand, with the present approach one can avoid the cost altogether by directly using the ufunc...

Overall, my sense remains to just be sure that the implementation matches the documentation, which states
```
    Equivalent to but faster than ``np.maximum(a_min, np.minimum(a, a_max))`
```
(which is consistent with the example of @tylerjereddy...), i.e., all we do is add a note that no check on `a_min <= a_max` is done.

tylerjereddy replied:
The [devdocs for `np.clip` ](https://www.numpy.org/devdocs/reference/generated/numpy.clip.html#numpy.clip) don't seem to have that statement, and that's what I wrote the property-based tests on, where the `Returns` entry indicates:

> An array with the elements of a, but where values < a_min are replaced with a_min, and those > a_max with a_max.

So, the proposal here is to amend the `numpy.clip` docstring with the pertinent content from `numpy.core.umath` in `ufunc_docstrings.py`, rather than take a performance hit & change behavior for the (admittedly bizarre) use case?

mhvk replied:
Good point...  I guess for the devdocs, one would have to clarify the order in which the replacement is done (as written, it is not quite consistent with what you find).  But, yes, my suggestion is indeed to ensure the documentation is correct also for the corner case.

eric-wieser replied:
Let's also check that I didn't accidentally swap the behavior from what it was before

tylerjereddy replied:
I went with "define and document" here for now, along with adjusting the cognate unit test to enforce the equivalence condition Marten dug up from the ufunc docs.

I'll mark this specific case as tentatively resolved to reduce review confusion.

tylerjereddy commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -2528,6 +2534,46 @@ def add_newdoc(place, name, doc):
+    is specified, values smaller than 0 become 0, and values larger
+    than 1 become 1.
+
+    Equivalent to but faster than ``np.maximum(a_min, np.minimum(a, a_max))``.
```
The [updated property-based test](https://github.com/tylerjereddy/numpy-property/blob/master/tests/test_clip.py#L12) still doesn't agree that this condition is enforced on master or this feature branch.

On feature branch: `Falsifying example: test_clip_scalar_ints(arr=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), amin=0, amax=-9223372036854775809)`

On master branch: `Falsifying example: test_clip_scalar_ints(arr=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), amin=0, amax=-1)`

tylerjereddy replied:
This was because `npy_ObjectClip` had the operations & operands jumbled around, so that it wasn't a true match to `np.maximum(amin, np.minimum(arr, amax))`.

I've added a unit test for the case used by `hypothesis` along with the patch to fix this.

tylerjereddy commented:
Down to just 1 of the 3 property-based tests failing, with: 

`Falsifying example: test_clip_mixed_arr_scalar(arr=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), amin=nan, amax=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32))`

tylerjereddy commented:
`RuntimeWarning: invalid value encountered in matmul` 32-bit failure is not related, but is observed from time to time in our CI...

tylerjereddy commented on numpy/core/src/umath/clip.c.src:
```diff
@@ -0,0 +1,119 @@
+/**begin repeat
+ * #name = FLOAT, DOUBLE, LONGDOUBLE#
+ */
+#define _NPY_@name@_MIN(a, b) (npy_isnan(a) ? (a) : PyArray_MIN(a, b))
```
My problem here is that npy_isnan(a) doesn't seem to work for "scalar" `np.nan` as a clip limit, so you don't actually propagate.

It looks like some thought has gone into this for this PR, with i.e., `_clip_dep_is_scalar_nan` and a `DeprecationWarning`, but the propagation definitely is not happening through the clip inner loop clip calls through MAX at the moment.

This is probably because there seems to be an interception with an infinite value replacing the scalar nan. I'm not sure how to proceed on that--it seems to be a discouraged use case, but not removed yet & I'm not sure if scalar nan should be truly exempted from propagation until the deprecation cycle is over?

tylerjereddy commented on numpy/core/_methods.py:
```diff
@@ -55,6 +57,80 @@ def _count_reduce_items(arr, axis):
+            warnings.warn(
+                "Passing `np.nan` to mean no clipping in np.clip has always "
+                "been unreliable, and is now deprecated. "
+                "In future, this will always return nan, like it already does "
```
The infinity substitutions above on scalar `np.nan` clip limits would seem to preclude the `npy_isnan()` checks in `clip.c.src` MIN and MAX preprocessor directives aimed at propagating nans.

eric-wieser replied:
Correct - but these substitutions only occur in the scalar case, not the array case. Unfortunately this was explicitly tested-for behavior, so I felt I had to preserve it.

tylerjereddy commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1775,11 +1798,94 @@ def test_clip_func_takes_out(self):
+    def test_NaT_propagation(self, arr, amin, amax):
+        # NOTE: the expected function spec doesn't
+        # propagate NaT, but clip() now does
+        expected = np.maximum(amin, np.minimum(arr, amax))
```
I reversed the MAX behavior for timedelta64 to propagate `NaT` as requested, but once again the fidelity with `np.maximum(amin, np.minimum(arr, amax))` is poor here--there's no propagation for this test case, so I've marked the unit test with xfail for now.

eric-wieser replied:
Hmm, that worries me - am I right in thinking that `np.maximum(datetime, datetime)` does not propagate `NaT`?

seberg replied:
Yeah, when Tyler looked at the NaT sorting I noticed this as well. Not sure I opened an issue for it, will check.

seberg replied:
Ah, Tyler did: gh-12797

tylerjereddy commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1775,11 +1798,94 @@ def test_clip_func_takes_out(self):
+        actual = np.clip(arr, amin, amax)
+        assert_equal(actual, exp)
+
+    @pytest.mark.xfail(reason="no scalar nan propagation yet")
```
need to decide if scalar `np.nan` should propagate so marking this test with xfail too for now; the function spec does propagate the `nan` here.

Historical behavior is yet another dimension I haven't fully explored, though the deprecation message implies propagation of nan has been the norm?

eric-wieser replied:
Historically Nan scalars propagate only in non-native memory orders, which I've deprecated

seberg replied:
We talked about this briefly yesterday. And I think the consensus is, that whatever the current discussion/state is exactly, we should be be practical. Especially how incorrectly sorted min and max value for clipping behaves is OK to be undefined (or raise an error). And it is OK to risk breaking current code.
With respect to NaN, I think it does not matter much which way we go and we should just decide for one thing. Although, in the case of NaN, I guess it may make sense to do a deprecation for the scalar case.
So the gist is: Don't worry too much about some tiny inconsistencies in corner cases of corner cases, it is fine to keep some things undefined (we could even make it well defined later if necessary).

eric-wieser replied:
> Although, in the case of NaN, I guess it may make sense to do a deprecation for the scalar case.

This is what I've implemented, simply because we actually had tests requiring that behavior

tylerjereddy commented:
Part of the reason this is stuck is because a number of core devs have expressed verbal favor for strictly enforcing `amin < amax`, including breaking backwards compatibility on the "abuse" where `amin` is larger, while the review in this PR currently proposes a new function spec that does not require that check.

The base argument is correctness over performance.

Is this still a point of contention?

mhvk commented:
Not just correctness - ufuncs generally do no argument checking - all they do is to catch errors. This would be like checking that arguments for `log` are positive.

My problem with sacrificing speed for making things fool-proof is that there is no way to avoid that penalty, while it is not particularly difficult to add tests outside of numpy.

Anyway, this discussion is probably best had outside of this PR. Since this PR does not change behaviour, it would seem reasonable to merge it.

eric-wieser commented:
I sort of dropped the ball on this one - I'll try to look at this in the next few days

seberg commented:
I would be happy to have errors on incorrect amax/amin values. Depending on changed behaviour, I think it may be a good idea for the scalar case (so before the dispatch to ufunc). I think we should be pragmatic and it does seem OK to me to just document it as "undefined", if future us think that we should enforce an error it would not be a big issue to go that way later, I think.

eric-wieser commented:
I don't think we need to worry about the checking in this PR. Let's leave things as they were before, and think about introducing that later.

tylerjereddy commented:
I'm showing [divergence](https://github.com/tylerjereddy/numpy-property) with current master for some cases, on the assumption the target behavior is `np.maximum(amin, np.minimum(arr, amax))`:

Fails on this branch:
`Falsifying example: test_clip_mixed_arr_scalar(arr=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), amin=nan, amax=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32))`
Fails on master branch:
`Falsifying example: test_clip_scalar_ints(arr=array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1]), amin=1, amax=0)`
`Falsifying example: test_clip_timedelta64(arr=array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1]), amin=1, amax=0)`

You're ok with those divergences? Even if the actual expected values are not set in stone for every single case just yet, it is the difference between branches I'm looking at now since the PR is effectively framed as preserving previous behavior.

mhvk commented:
I think the first failure is expected - the PR still treats `nan` as "ignore" - the others confuse me - we should definitely ensure it doesn't depend on type!

tylerjereddy commented:
Let's start with this case:

```python
result = np.clip(np.ones(10, dtype=np.int64), 1, 0)
print("result:", result)
print("result.dtype:", result.dtype)
# on master: [0 0 0 0 0 0 0 0 0 0]
# on master: int64

# on feature branch: [1 1 1 1 1 1 1 1 1 1]
# on feature branch: int64
```

The feature branch wins by matching `expected = np.maximum(amin, np.minimum(arr, amax))`, but it is a break on strictly claiming there's no behavior change from master.

That said, this is a relatively pathological scenario cooked up by a library designed to find such scenarios, so you may be just fine with that?

eric-wieser commented:
Does swapping the order of min and max in the expectation and this branch fix things?

I honestly did not consider this case when implementing. Given the goal here was not to fix the weirdness but move to ufuncs, I think we should try and leave things exactly as they were wherever possible.

rgommers commented:
> I think we should try and leave things exactly as they were wherever possible.

A short comment, given that I voiced an opinion in the call but didn't look at the code in detail:
- leaving current weird behavior unchanged is of course fine if that is easier and makes that this can be merged sooner
- on the other hand if it's less work to go to more correct behavior, just consider the current behavior a bug and do it. `amin > amax` clearly was never well thought out and should either be undefined behavior or (preferably) raise a `ValueError`
- formalizing weird behavior in an explicit "spec" is not right. if you're ignoring it, just leave as is and leave the option open to fix that in the future

eric-wieser commented:
@tylerjereddy: here's where the weirdness comes from (on master):
```
In [5]: np.ones(10, dtype=object).clip(1, 0)
Out[5]: array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=object)

In [6]: np.ones(10, dtype=np.int64).clip(1, 0)
Out[6]: array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)
```
I copied the implementation from the object loop, but it looks like it varies by type. It's probably best to try and preserve the old behavior for the builtin numeric types, and break the object array behavior.

So in essence, swap the order of min and max in all of the ufunc loops.

tylerjereddy commented:
> So in essence, swap the order of min and max in all of the ufunc loops.

That didn't quite work in my hands, but maybe I'm not understanding. The simplest I could come up with was using an outer Min when `min > max`.

I added a new test `test_ones_pathological()` based on the most recent discussion above. In a temporarily separate commit I've added the draft changes needed for that new test, and some of the previous tests, to pass (though CI already disagrees with local suite).

It will need revision / less code duplication, even if we use this approach for the loops. I think I got a little stuck because I didn't have immediate access to `>` operand for `cfloat` and related types, and ended up with some inner loop duplication. Could also maybe just use special names for `_NPY_CLIP` loops by type if needed I guess. The changes seem a little gruesome though!

eric-wieser commented:
Your last commit doesn't look right to me. Do you mind if I blow it away?

eric-wieser committed:
BUG: Restore the old non-object behavior for min > max

This changes the object array behavior to match the other behavior

eric-wieser commented:
Updated. I rewrote the first commit slightly to pull in the obvious bug fixes, but left the rest of your fixup.

tylerjereddy commented:
Can you confirm that these are now expected to fail on master & pass on this feature branch, as far as conforming to our new "function spec" with the flip-around on min & max:

```
test_clip_array_floats(arr=array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
                                  amin=array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]), 
                                  amax=array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]))        
test_clip_scalar_ints(arr=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), amin=1, amax=0)
```

Those are the residual differences in the property tests after most recent push.

eric-wieser commented:
Where is `test_clip_array_floats` defined?

tylerjereddy commented:
Both produce all `1` instead of all `0` arrays on master & are defined [here](https://github.com/tylerjereddy/numpy-property/blob/master/tests/test_clip.py)

eric-wieser commented:
I'm super confused, that sounds like the exact opposite to what I observed the other day. Will check my scrollback tonight

tylerjereddy commented:
Yeah, it is really confusing. I still don't get your basis for the new function spec in [this comment of yours](https://github.com/numpy/numpy/pull/12519#issuecomment-482014395):

Check the comment in the second example there, which starts with x as all ones
`min(max(x, amax), amin)` 
min(max(1, 0), 1) -> 1 (master never did this for non-object types, so why was this the spec?)

A local reproducer independent of the third-party testing library:

```python
# require recent master
assert 'c73865' in np.__version__

arr0 = np.zeros(10, dtype=np.int64)
arr1 = np.ones(10, dtype=np.int64)
for arr in [arr0, arr1]:
    result = np.clip(arr, 1, 0)
    print("result:", result)
    print("result.dtype:", result.dtype)
```

```
result: [1 1 1 1 1 1 1 1 1 1]
result.dtype: int64
result: [0 0 0 0 0 0 0 0 0 0]
result.dtype: int64
```

So, the new spec of `min(max(x, max), min)` would do the opposite of what master does above producing:
```
min(max(x, amax), amin)
min(max(x, 0), 1) # substitute clip limits
min(max(0, 0), 1) -> 0 # start from zeros would go to 0, but master goes to 1
min(max(1, 0), 1) -> 1 # start from ones would go to 1, but master goes to zero, just as in your original example
```

eric-wieser commented:
Testing on 1.15.4:
```
for t in '?bhilqpBHILQPefdgFDGO':
	t = np.dtype(t)
	print(t, np.ones(1, dtype=t).clip(t.type(1), t.type(0)))
```
gives
```
bool [0]
int8 [0]
int16 [0]
int32 [0]
int32 [0]
int64 [0]
int64 [0]
uint8 [0]
uint16 [0]
uint32 [0]
uint32 [0]
uint64 [0.]
uint64 [0.]
float16 [0.]
float32 [0.]
float64 [0.]
float64 [0.]
complex64 [0.+0.j]
complex128 [0.+0.j]
complex128 [0.+0.j]
object [1]
```

eric-wieser commented:
> Both produce all 1 instead of all 0 arrays on master

I've realized I do not understand this statement. Are you saying they produce 1 arrays or 0 arrays on master?

eric-wieser commented:
Alright, here's where the confusion comes from. You keep implying things are backwards, so we keep swapping them and making them backwards somewhere else:
```
>>> np.clip([1], 1, 0)
array([0])
>>> np.clip([1], [1], [0])
array([1])
```

tylerjereddy commented:
I think I'm just looking for the property-based tests to give the same results on master and feature, except for those cases we've made a conscious decision that they should differ. But as Ralf noted, this may be a little bit nutty to attempt for the amin > amax case.

eric-wieser commented:
To expand on my somewhat cryptic comment above, `clip(one_x, one_min, zero_max)` gives on master:

* `1` if `one_x.dtype == object`
* `1` if `one_min.ndim > 0  or zero_max.ndim > 0`
* `0` otherwise (scalar and non-object)

It would be bizarre to keep the second bullet as a special case, ufunc behavior should not depend on scalars. From that perspective, we have three choices:

1. Always return `0`
   * keeps behavior of the most common scalar case
1. Return `0` for all dtypes except `object`
   * keeps behavior of all (non-nan) scalar cases,
   * maintains weird special casing
1. Always return `1`
   * keeps all array behavior

My leaning is towards \#1, since I think calling `clip` with scalars is by far the most common.  I think that's what my latest commit implements.

Should we confirm that with the mailing list, or just write it off as undefined behavior and go ahead?

mhvk commented:
I'm in favour of your option 1 - it should always be consistent, and I think processing arguments "in order" is most logical, i.e., clipping to the maximum is done last. I don't think this has to go by the mailing list - I don't think anybody can reasonably expect certain behaviour, certainly when it was somewhat random so far.

eric-wieser commented:
Found another behavior change.

Before:
```python
>>>  np.clip(np.uint64([3]), -1, 20)
array([3], dtype=uint64)
>>>  np.clip(np.uint64([3]), -1, [20, 20])
array([3., 3.])
>>> np.clip(np.uint16(3), np.int8(-1), np.int8(10)).dtype
dtype('uint16')
```
After:
```python
>>>  np.clip(np.uint64([3]), -1, 20)
array([3.])  # change
>>>  np.clip(np.uint64([3]), -1, [20, 20])
array([3., 3.])  # matches
>>> np.clip(np.uint16(3), np.int8(-1), np.int8(10)).dtype
dtype('int32')  # change
```

eric-wieser commented:
@mhvk: Do you think the casting weirdness above matters?

I've spent the last hour trying to come up with a way to detect and deprecate it, but our casting mechanism is rather rigid, and doesn't provide enough introspection for me to do anything useful.

I don't think it's worth investing much more time in trying to maintain weirdness stemming from a poorly handled optimization (the clip fast-path)

mhvk commented:
@eric-wieser - no, I don't think we have to be that precise in preserving behaviour from the fast path. Indeed, I think you already go out of your way to preserve the `nan` behaviour.

eric-wieser commented:
In that case, what do you think the way forward is here?

mhvk commented:
I'm quite happy to merge... @tylerjereddy?

charris commented:
With the various fixes/deprecations, this needs a release note.

tylerjereddy commented:
If I remove `nan` scenarios, then the property-based tests pass on this feature branch for matching the spec `np.minimum(amax, np.maximum(arr, amin))`

master branch struggles a bit for the above spec when amin >= amax.

Anyway, I think most other core devs I've discussed this with don't really want to spend energy on the pathological cases, and the error handling there is another discussion re: performance / interface to ufuncs.

One of the tests I added has an xfail--not sure if want to keep that one.

tylerjereddy commented:
Just needs:
- release note
- rebase / force-push

If CI all-green after that then I merge based on core dev +1 from Marten and my thorough review, etc.

eric-wieser committed:
DOC: Add release note

charris commented:
Needs rebase. Test failure is fixed in master.

charris commented:
Needs rebase. Test error is fixed in master.

tylerjereddy commented:
rebased / force pushed -- if CI is all green I plan to merge as noted above

tylerjereddy commented:
The CI failures are from things like `numpy/core/src/umath/simd.inc.src:1407:1: warning: ‘AVX512F_log_FLOAT’ defined but not used [-Wunused-function]`.

This is presumably coming in from work by @r-devulap and @juliantaylor?

Hopefully no merge conflict in release notes arises while waiting for this to get resolved / rebased.

r-devulap commented:
I think the reason you are seeing that error is the file `simd.inc` is included in `fast_loop_macros.h` which is then included in `clip.c.src`. `simd.inc ` contains the code for static functions `AVX2_log_FLOAT ` etc which ends up not being used anywhere in `clip.c.src` and hence the error. inlining the ISA_exp/log_FLOAT functions with NPY_INLINE keyword solves the build issue you are seeing.

r-devulap commented:
Is there a reason why `simd.inc` is included in `fast_loop_macros.h`?

tylerjereddy commented:
Perhaps we can make that refinement if the inclusion is not needed @eric-wieser ?

r-devulap commented:
I was able to build without `#include<simd.inc>` in `fast_loop_macros.h`, so I am not sure if that include is needed.

tylerjereddy commented:
Thanks, that's helpful feedback! Eric or Marten can probably confirm the safety of removal, but sounds straightforward.

eric-wieser commented:
I was under the impression that some of the macros in that file expand to things from the simd header.

I think adding inline is the right way to go here.

tylerjereddy committed:
MAINT: reviewer adjustments re: AVX

* ISA_exp/log have been inlined to prevent
build errors when included in PR 12519

tylerjereddy commented:
Ok, I inlined the low-level exp / log stuff & local test suite results are unchanged while the build warnings-turned-errors disappear. Hopefully CI agrees.

tylerjereddy commented:
Thanks Eric and all reviewers. Merging with fingers crossed based on various discussions above.
